VariantShredding.md

```diff
@@ -25,276 +25,302 @@
+Each shredded field in the `typed_value` group is represented as a required group that contains optional `value` and `typed_value` fields.
+The `value` field stores the value as Variant-encoded `binary` when the `typed_value` cannot represent the field.
+This layout enables readers to skip data based on the field statistics for `value` and `typed_value`.
 
+For example, a Variant `event` field may shred `event_type` (`string`) and `event_ts` (`timestamp`) columns using the following definition:
 ```
-# Constructs a Variant from `variant_value`, `object`, `array` and `typed_value`.
-# Only one of object, array and typed_value may be non-null.
-def ConstructVariant(variant_value, object, array, typed_value):
-  if object is null and array is null and typed_value is null and variant_value is null: return VariantNull 
-  if object is not null:
-    return ConstructObject(variant_value, object)
-  elif array is not null:
-    return ConstructArray(array)
-  elif typed_value is not null:
-    return cast(typed_value as Variant)
-  else:
-    variant_value
-
-# Construct an object from an `object` group, and a (possibly null) Variant variant_value
-def ConstructObject(variant_value, object):
-  # If variant_value is present and is not an Object, then the result is ambiguous.
-  assert(variant_value is null or is_object(variant_value))
-  # Null fields in the object are missing from the reconstructed Variant.
-  nonnull_object_fields = object.fields.filter(field -> field is not null)
-  all_keys = Union(variant_value.keys, non_null_object_fields)
-  return VariantObject(all_keys.map { key ->
-    if key in object: (key, ConstructVariant(object[key].variant_value, object[key].object, object[key].array, object[key].typed_value))
-    else: (key, variant_value[key])
-  })
-
-def ConstructArray(array):
-  newVariantArray = VariantArray()
-  for i in range(array.size):
-    newVariantArray.append(ConstructVariant(array[i].variant_value, array[i].object, array[i].array, array[i].typed_value)
+optional group event (VARIANT) {
+  required binary metadata;
+  optional binary value;                # a variant, expected to be an object
+  optional group typed_value {          # shredded fields for the variant object
+    required group event_type {         # shredded field for event_type
+      optional binary value;
+      optional binary typed_value (STRING);
+    }
+    required group event_ts {           # shredded field for event_ts
+      optional binary value;
+      optional int64 typed_value (TIMESTAMP(true, MICROS));
+    }
+  }
+}
 ```
 
-# Nested Parquet Example
+The group for each named field must be required.
 
-This section describes a more deeply nested example, using a top-level array as the shredding type.
+A field's `value` and `typed_value` are set to null (missing) to indicate that the field does not exist in the variant.
```

Comment:
That would have some inconsistency with the encoding for primitive values that both nulls are invalid. 
Why do we make the group required? Can we have group optional and then the group is not present, then that means the field is missing.

Reply:
This is consistent. The table above states that if both `value` and `typed_value` are null, then the value is missing.

Reply:
> What we would end up needing to say here is that when the field's group is null OR when it is non-null but both value and typed_value are null, the value is missing.

Doesn't the field's group being null imply both leaf fields are null though (so it can be interpreted as always missing)? It also seems like this case needs to happen anyways for any nested shredded objects?

I have to think more about if I feel strongly about it but if you assume that data is read into a sparse memory model (nulls values do not take space, e.g. Arrow but where null subfields don't have a slot associated with them) before reconstructing a variant it is cheaper to know that both fields are null at the upper level, so you don't take more space at the leaves (e.g. you could save 2 bits per missing value to void marking both as null).

Reply:
> Doesn't the field's group being null imply both leaf fields are null though

@emkornfield, it would but the point is that we don't need extra ways to express this so the field group should be `required` and not `optional`. Making the group optional introduces more cases that need to be documented, but don't provide any additional capability.

Reply:
I think leaving as is is fine, we can always iterate if there is a strong reason do so.  @aihuaxu do you have further concerns here?